play-framework/async-http/test/controllers/HomeControllerTest.java

```diff
@@ -0,0 +1,246 @@
+import play.inject.guice.GuiceApplicationBuilder;
+import play.libs.concurrent.Futures;
+import play.libs.ws.WSClient;
+import play.libs.ws.WSResponse;
+import play.libs.ws.ahc.AhcCurlRequestLogger;
+import play.mvc.Result;
+import play.mvc.Results;
+import play.test.WithServer;
+
+public class HomeControllerTest extends WithServer {
+
+    private final Logger log = LoggerFactory.getLogger(HomeControllerTest.class);
+    private String url;
+    private int port;
+
+    @Override
+    protected Application provideApplication() {
+        return new GuiceApplicationBuilder().build();
+    }
+
+    @Before
+    public void setup() {
+        OptionalInt optHttpsPort = testServer.getRunningHttpsPort();
+        if (optHttpsPort.isPresent()) {
+            port = optHttpsPort.getAsInt();
+            url = "https://localhost:" + port;
+        } else {
+            port = testServer.getRunningHttpPort()
+                             .getAsInt();
+            url = "http://localhost:" + port;
+        }
+    }
+
+    @Test
+    public void givenASingleGetRequestWhenResponseThenBlockWithCompletableAndLog()
+      throws InterruptedException, ExecutionException {
+        WSClient ws = play.test.WSTestClient.newClient(port);
+        WSResponse wsResponse = ws.url(url)
+                                  .setRequestFilter(new AhcCurlRequestLogger())
+                                  .addHeader("key", "value")
+                                  .addQueryParameter("num", "" + 1)
+                                  .get()
+                                  .toCompletableFuture()
+                                  .get();
+
+        log.debug("Thread#" + Thread.currentThread()
+                                    .getId() + " Request complete: Response code = "
+          + wsResponse.getStatus()
+          + " | Response: " + wsResponse.getBody() + " | Current Time:"
+          + System.currentTimeMillis());
+    }
+
+    @Test
+    public void givenASingleGetRequestWhenResponseThenLog() throws Exception {
+        AtomicInteger completedReqs = new AtomicInteger(0);
+        CountDownLatch latch = new CountDownLatch(1);
+
+        WSClient ws = play.test.WSTestClient.newClient(port);
+        ws.url(url)
+          .setRequestFilter(new AhcCurlRequestLogger())
+          .addHeader("key", "value")
+          .addQueryParameter("num", "" + 1)
+          .get()
+          .thenAccept(r -> {
+              log.debug("Thread#" + Thread.currentThread()
+                                          .getId() + " Request complete: Response code = "
+                + r.getStatus()
+                + " | Response: " + r.getBody() + " | Current Time:" + System.currentTimeMillis());
+              latch.countDown();
+              completedReqs.incrementAndGet();
+          });
+
+        log.debug(
+          "Waiting for requests to be completed. Current Time: " + System.currentTimeMillis());
+        latch.await();
+        assertEquals(1, completedReqs.get());
```

Comment:
The `latch` and the `completedReqs` are doing the same job. You only need one. You may be able to put a timeout on the `await` and then check if the `getCount` on the latch to make sure you didn't get to the `assertEquals` because of timeout, rather than the latch hitting `0`.

Reply:
Good idea. I never used latch so I didn't know about the getCount.
I removed the AtomicInt and used the latch.getCount instead.